sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala

```diff
@@ -91,7 +91,7 @@ trait CaseWhenLike extends Expression {
 
   // both then and else expressions should be considered.
   def valueTypes: Seq[DataType] = (thenList ++ elseValue).map(_.dataType)
-  def valueTypesEqual: Boolean = valueTypes.distinct.size == 1
+  def valueTypesEqual: Boolean = valueTypes.map(_.asNullable).distinct.size == 1
```

Comment:
I feel it is not the right fix. We cannot just cast them all to nullable. I feel the best fix should be when we try to find the common type, we understand non-nullable field can be cast to nullable field.